src/main.cpp

```diff
@@ -6259,7 +6273,12 @@ bool static ProcessMessage(const CChainParams& chainparams, CNode* pfrom, string
         if (pindexLast)
             UpdateBlockAvailability(pfrom->GetId(), pindexLast->GetBlockHash());
 
-        if (nCount == MAX_HEADERS_RESULTS && pindexLast) {
+        // Temporary, until we're sure the optimization works
+        if (nCount == MAX_HEADERS_RESULTS && pindexLast && !hasNewHeaders) {
+            LogPrint("net", "NO more getheaders (%d) to send to peer=%d (startheight:%d)\n", pindexLast->nHeight, pfrom->id, pfrom->nStartingHeight);
+        }
```

Comment:
This `if` could be combined with the next one, but this `LogPrint()` probably should be removed once we're sure this works, and it's easier to just remove this `if`.